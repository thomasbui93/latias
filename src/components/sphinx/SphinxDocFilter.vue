<template>
  <div class="sphinx-search">
    <div class="sphinx-search__field">
      <input class="input" type="text" v-model="text" placeholder="What are you searching for?">
    </div>
    <div class="sphinx-search__field">
      <div class="select">
        <select v-model="sortType">
          <option value="ASC">A-Z</option>
          <option value="DESC">Z-A</option>
        </select>
      </div>
      <div class="select">
        <select v-model="sortBy">
          <option value="createdAt">Date</option>
          <option value="question">Question</option>
          <option value="level">Level</option>
        </select>
      </div>
    </div>
    <div class="sphinx-search__field">
      <button class="button is-primary" @click="triggerChange">Search</button>
    </div>
  </div>
</template>

<script>
export default {
  props: {
    value: {
      type: Object,
      default: () => ({
        text: '',
        keywords: [],
        sortType: 'ASC',
        sortBy: 'createdAt',
      }),
      validator(data) {
        return ['ASC', 'DESC'].indexOf(data.sortType) > -1
          && ['createdAt', 'question', 'level'].indexOf(data.sortBy) > -1;
      },
    },
  },
  computed: {
    keywordList() {
      return this.keywords.split(',');
    },
  },
  data() {
    const {
      text, keywords, sortType, sortBy,
    } = this.value;
    return {
      text,
      keywords,
      sortType,
      sortBy,
    };
  },
  methods: {
    triggerChange() {
      this.$emit('input', {
        text: this.text,
        keywords: this.keywords,
        sortType: this.sortType,
        sortBy: this.sortBy,
      });
    },
  },
};
</script>
