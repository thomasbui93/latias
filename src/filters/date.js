import Vue from 'vue';
import dayjs from 'dayjs';

export const normalizedDate = (value) => {
  if (!value) return '';
  const date = dayjs(value);
  return date.isValid() ? dayjs(value).format('YYYY-MM-DD') : '';
};

export default Vue.filter('dateNormalized', normalizedDate);
