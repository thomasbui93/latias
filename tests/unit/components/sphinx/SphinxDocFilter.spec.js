import SphinxDocFilter from '@/components/sphinx/SphinxDocFilter.vue';
import { mount, shallowMount } from '@vue/test-utils';
import '@/filters/date';

describe('SphinxDocFilter component', () => {
  it('match snapshot', () => {
    const wrapper = shallowMount(SphinxDocFilter, {
      propsData: {
        value: {
          text: 'sample',
          keywords: ['keyword1', 'keyword2'],
          sortType: 'ASC',
          sortBy: 'createdAt',
          level: 'ADVANCED',
        },
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('render default value if prop value is missing.', () => {
    const wrapper = mount(SphinxDocFilter);
    expect(wrapper.vm.value).toEqual({
      text: '',
      keywords: [],
      sortType: 'ASC',
      sortBy: 'createdAt',
      level: '',
    });
  });

  it('click on level to emit toggleLevel', () => {
    const wrapper = mount(SphinxDocFilter, {
      propsData: {
        value: {
          text: 'sample',
          keywords: ['keyword1', 'keyword2'],
          sortType: 'ASC',
          sortBy: 'createdAt',
          level: 'ADVANCED',
        },
      },
    });
    wrapper.find('.sphinx-level').trigger('click');
    expect(wrapper.emitted().toggleLevel[0]).toEqual(['ADVANCED']);
  });

  it('click on keyword to emit toggleKeyword', () => {
    const wrapper = mount(SphinxDocFilter, {
      propsData: {
        value: {
          text: 'sample',
          keywords: ['keyword1', 'keyword2'],
          sortType: 'ASC',
          sortBy: 'createdAt',
          level: 'ADVANCED',
        },
      },
    });
    wrapper.find('.sphinx-tag:nth-child(2)').trigger('click');
    expect(wrapper.emitted().toggleKeyword[0]).toEqual(['keyword2']);
  });

  it('click on button to emit change', () => {
    const wrapper = mount(SphinxDocFilter, {
      propsData: {
        value: {
          text: 'sample',
          keywords: ['keyword1', 'keyword2'],
          sortType: 'ASC',
          sortBy: 'createdAt',
          level: 'ADVANCED',
        },
      },
    });
    wrapper.find('.sphinx-search__field-action button').trigger('click');
    expect(wrapper.emitted().input[0]).toEqual([{
      text: wrapper.vm.text,
      keywords: wrapper.vm.keywords,
      sortType: wrapper.vm.sortType,
      sortBy: wrapper.vm.sortBy,
      level: wrapper.vm.value.level,
    }]);
  });
});
