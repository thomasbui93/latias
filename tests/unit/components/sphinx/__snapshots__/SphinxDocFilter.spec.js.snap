// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SphinxDocFilter component match snapshot 1`] = `
<div class="sphinx-search">
  <div class="sphinx-search__field-group">
    <div class="sphinx-search__field"><input type="text" placeholder="What are you searching for?" class="input"></div>
    <div class="sphinx-search__field">
      <div class="select"><select>
          <option value="ASC">A-Z</option>
          <option value="DESC">Z-A</option>
        </select></div>
      <div class="select"><select>
          <option value="createdAt">Date</option>
          <option value="question">Question</option>
          <option value="level">Level</option>
        </select></div>
    </div>
  </div>
  <div class="sphinx-search__field">
    <div class="sphinx-tag">
      keyword1
      <span class="sphinx-tag__close">-</span></div>
    <div class="sphinx-tag">
      keyword2
      <span class="sphinx-tag__close">-</span></div> <span class="sphinx-tag sphinx-level">
      ADVANCED
    </span>
  </div>
  <div class="sphinx-search__field-action"><button class="button is-primary">Search</button></div>
</div>
`;
